File: kmer_filter.cc
#include <jellyfish/jellyfish.hpp>
#include <iostream>
using namespace std;

int main(int argc, char const *argv[])
{
    std::ifstream ifs("/home/user/duanran/repo/spider-web/ref/NC_010079.k32", ios::binary);
    std::ofstream ofs("/home/user/duanran/repo/spider-web/ref/NC_010079.k32.filtered.fa", ios::binary);

    jellyfish::file_header header(ifs);
    jellyfish::mer_dna::k(header.key_len() / 2);
    binary_reader reader(ifs, &header);

    text_writer writer;
    for (int i = 0; reader.next(); i ++) {
        ofs << ">" << reader.val() << "\n" << reader.key() << "\n";
    }

    ifs.close();
    ofs.close();
    return 0;
}

File: kmer_filter.py
import json
from collections import Counter

# jtk: jellyfish kmer count result with '--text' flag set
def jtk_iterator(path):
    f = open(path)
    header_size = int(f.read(9)) # 读取接下来的json大小
    header_s = f.read(header_size).replace('\x00', '') # 读取json，其末端可能有\x00填充，需要切掉
    header = json.loads(header_s)
    while True:
        s = f.readline()
        if s == "":
            break
        else:
            pair = s.strip().split()
            yield pair[0], int(pair[1])
    f.close()

jtk_path = "/home/user/duanran/repo/spider-web/ref/NC_008253.jtk"
fk_path = "/home/user/duanran/repo/spider-web/ref/NC_008253.filtered.fk"
of = open(fk_path, "w")
jtk = jtk_iterator(jtk_path)
for key, val in jtk:
    c = Counter(key)
    proportion = (c['C'] + c['G']) / len(key)
    
